src/commandSocket.cpp

```diff
@@ -57,93 +57,135 @@ class Listener : public Arcus::SocketListener
 
+    std::shared_ptr<cura::proto::LayerOptimized> getOptimizedLayerById(int id);
+
     Arcus::Socket* socket;
     
     // Number of objects that need to be sliced
     int object_count;
 
-    // Number of sliced objects for this sliced object list
-    int sliced_objects;
-
-    // Number of layers sent to the front end so far
-    // Used for incrementing the current layer in one at a time mode
-    int current_layer_count;
-    int current_layer_offset;
-
     std::string temp_gcode_file;
     std::ostringstream gcode_output_stream;
     
     // Print object that olds one or more meshes that need to be sliced. 
     std::vector< std::shared_ptr<MeshGroup> > objects_to_slice;
 
-    std::unordered_map<int, std::shared_ptr<cura::proto::Layer>> sliced_layers;
+    SliceDataStruct<cura::proto::Layer> sliced_layers;
+    SliceDataStruct<cura::proto::LayerOptimized> optimized_layers;
 };
 
-class CommandSocket::PolygonCompiler
+class CommandSocket::PathCompiler
 {
-	static_assert(sizeof(PrintFeatureType) == 1, "To be compatible with the Cura frontend code PrintFeatureType needs to be of size 1");
-	CommandSocket::Private& _cs_private_data;
-	int _layer_nr;
-	std::vector<PrintFeatureType> _line_types;
-	std::vector<int> _line_widths;
-	cura::Polygon _poly;
-
-	PolygonCompiler(const PolygonCompiler&) = delete;
-	PolygonCompiler& operator=(const PolygonCompiler&) = delete;
+    static_assert(sizeof(PrintFeatureType) == 1, "To be compatible with the Cura frontend code PrintFeatureType needs to be of size 1");
+    CommandSocket::Private& _cs_private_data;
+    int _layer_nr;
+    std::vector<PrintFeatureType> line_types;
+    std::vector<int> line_widths;
+    std::vector<Point> points;
+
+    PathCompiler(const PathCompiler&) = delete;
+    PathCompiler& operator=(const PathCompiler&) = delete;
 public:
-	PolygonCompiler(CommandSocket::Private& cs_private_data):
-		_cs_private_data(cs_private_data),
-		_layer_nr( 0 ),
-		_line_widths(),
-		_poly()
-	{}
-	~PolygonCompiler(){
-		if( _poly.size() )
-			flushPolygon();
-	}
-
-	void setLayer(int new_layer_nr){
-		if(_layer_nr != new_layer_nr ){
-			flushPolygon();
-			_layer_nr = new_layer_nr;
-		}
-	}
-
-	void handleInitialPoint( Point from ){
-		if( _poly.size() == 0 ){
-			_poly.add( from );
-		}
-		else if( from != _poly.back() )
-		{
-			_poly.add( from );
-			_line_types.push_back( PrintFeatureType::NoneType );
-			_line_widths.push_back( 1.0 );
-		}
-	}
-
-	void flushPolygon();
-	void sendLine(PrintFeatureType print_feature_type, Point from, Point to, int line_width);
-    void sendPolygon(PrintFeatureType print_feature_type, Polygon poly, int line_width);
+    PathCompiler(CommandSocket::Private& cs_private_data):
+        _cs_private_data(cs_private_data),
+        _layer_nr( 0 ),
+        line_widths(),
+        points()
+    {}
+    ~PathCompiler()
+    {
+        if( points.size() )
+            flushPathSegments();
+    }
+
+    /*!
```

Comment:
I'd rather see the documentation with the declaration, rather than the definition.
However, I'm not blocking merging this PR on that..